api/v1alpha1/flowcollector_types.go

```diff
@@ -34,6 +34,13 @@ func (spec *FlowCollectorSpec) UseEBPF() bool  { return spec.Agent.Type == Agent
 func (spec *FlowCollectorSpec) UseIPFIX() bool { return spec.Agent.Type == AgentIPFIX }
 func (spec *FlowCollectorSpec) UseKafka() bool { return spec.DeploymentModel == DeploymentModelKafka }
 
+func (spec *FlowCollectorSpec) GetSampling() int {
+	if spec.UseEBPF() {
+		return int(*spec.Agent.EBPF.Sampling)
+	}
+	return int(spec.Agent.IPFIX.Sampling)
+}
```

Comment:
Until now, helper directly related to the API were added here. Similar to `UseEBPF`, `UseIPFIX`...

It does not need to be a method but it improve code readability IMO.

I count 9 helper function, do you think we should move all of them in a new helper package?

Reply:
I think so giving we will have different versions it will be a hassle, do we need to copy them to newer version or keep in the old one what if we need to delete the old api so additional Maintenance for no reason imo

Reply:
@msherif1234 Personally I've found these methods quite nice to use and I don't think I'm alone - but for sure they have to stay trivial being just syntactic sugar, directly tied to nothing but the model itself, IMO.

I don't really get the comment on the maintenance hassle: if we use helpers, that would be the same, or even worse, as they still need to be duplicated per crd version? When creating a new version, we just copy from the existing one, update the package name and that's it. Am I missing something here?

Reply:
+1 to move away functions from API. On top of that, we only use the 'latest' ones, not the previous ones

Reply:
@msherif1234 when creating a new API, you would basically start from the existing, so yes you copy-paste the whole file without any special attention for the syntax sugar. I don't see the maintenance hassle, am I missing it?

> what is the issue moving them out of the xxx_types.go to helper pkg ?

What annoys me here is not to choose one or another way of doing it. It's almost the same. I have my preference because it's just easier to write `myvar.` and get the IDE autocompletion than having to retrieve the helper package name (assuming we use a dedicated package for that, since catch-all helper packages are also code-smells). When writing code, it's more immediate to start from the variable itself. But, if the code was written the other way, I wouldn't care because this is really just a matter of code style with very low impact, I would consider it nitpicking and just a matter of personal preference. Certainly not asking the author of a PR who is adding an additional bit on top of that existing code, to do a refactoring for this, given the value added is quite debatable. I think we need to ack that code styles are full of nuances, it's hardly never a single way for doing things, and unless there is an obvious or well argumented benefit in forcing in a way or another (e.g: consistency, performance, maintainability ...) , IMO it's better to admit that every contributor has some degree of initiative in how they approach such code style.

So back to my original question, am I really missing something about the maintenance hassle, that would be a good reason for refactoring? Or am I wrong to say that these parts can simply be copy-pasted like the rest of the API when we introduce a new version?

Reply:
Just the cp and paste I learnt from api teams reviews to all previous apis to keep the types at the minimum since then keep applying this rules

Reply:
I pushed a new commit with the refactoring.

I have to say, I am not a big fan of the changes. IMO this make the code a bit less readable.

Reply:
what about embeding @OlivierCazade ? It's a bigger refactor so I suggest to do it in a followup but it will be transparent in terms of readability.
https://go.dev/doc/effective_go#embedding

```
package helper

import (
	flowslatest "github.com/netobserv/network-observability-operator/api/v1beta1"
)

type Desired struct {
	flowCollector flowslatest.FlowCollector
}

func (dc *Desired) GetSampling() int {
	if dc.UseEBPF() {
		return int(*dc.flowCollector.Spec.Agent.EBPF.Sampling)
	}
	return int(dc.flowCollector.Spec.Agent.IPFIX.Sampling)
}

func (dc *Desired) UseEBPF() bool {
	return dc.flowCollector.Spec.Agent.Type == flowslatest.AgentEBPF
}

...
```

And change `desired` type in `flowcollector_controller.go` Reconcile loop from `FlowCollector` to `Desired`

Reply:
@jpinsonneau this would work for the methods directly attached to the `FlowCollector` top structure, but this would not work for the methods attached to other sub structures.

Reply:
yes it needs a big refactor to rely on the `Desired` struct all the time. 
The good thing with that approach is that you have access to all informations everywhere so you will not have to refactor again in the future as you did in the PR.

Reply:
@jpinsonneau , I think Olivier's point is something like that: imagine I have a "UseMTLS" helper defined as:

```go
func (tls *ClientTLS) UseMTLS() bool { return tls.UserCert.Name != "" }
```

Since ClientTLS is a struct embedded inside FlowCollector and used in different places, with your proposal we would have to duplicate this helper for each embedded case, such as:

```go
func (dc *Desired) KafkaUseMTLS() bool {
	return dc.flowCollector.Spec.Kafka.TLS.UserCert.Name != ""
}

func (dc *Desired) LokiUseMTLS() bool {
	return dc.flowCollector.Spec.Loki.TLS.UserCert.Name != ""
}
```

Reply:
We already created [another helper](https://github.com/netobserv/network-observability-operator/blob/main/pkg/helper/certificates.go) relying on functions for that situation. I'm not saying we should have only one solution; just adapt according to the needs.

Reply:
If you guys are not confortable with [these changes](https://github.com/netobserv/network-observability-operator/pull/264/commits/69e9bd9cc78a9070de3a43a4ac95e1cbad1cc460), feel free to rollback and let's take more time to see the impacts.

Reply:
I think relying sometimes on external helpers and sometimes on a solution based on a `Desired` wrapper would only make things worse when writing code and looking to retrieve those helpers.

I'll give my +1 to the PR to not block it any further (too much time and energy spent here) but with the feeling that I don't think we've improved the code here

/lgtm

Reply:
Epilogue: the topic triggered some discussion with the API team, having functions in api was not explicitly prohibited, but they said they will add that to their [convention doc](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md). My understanding is that our functions are not fundamentally bad but they need clear guidelines in order to ease their review process so they will prohibit that. So, given we want to follow those conventions (as much as possible), that's a win for the helper approach.